src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesFromShardsDataflowTransform.java

```diff
@@ -0,0 +1,165 @@
+import org.broadinstitute.hellbender.engine.dataflow.DoFnWLog;
+import org.broadinstitute.hellbender.utils.dataflow.lcollections.LocalCollection;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.broadinstitute.hellbender.utils.read.markduplicates.OpticalDuplicateFinder;
+
+import java.util.ArrayList;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The main transform for MarkDuplicates. Takes reads, returns them with some reads marked as duplicates.
+ */
+public final class MarkDuplicatesFromShardsDataflowTransform extends PTransform<PCollection<KV<String, Iterable<GATKRead>>>, PCollection<GATKRead>> {
+    private static final long serialVersionUID = 1l;
+
+
+    /**
+     *  The header for all reads processed.
+     */
+    private final PCollectionView<SAMFileHeader> header;
+    /**
+     * Contains methods for finding optical duplicates.
+     */
+    private final PCollectionView<OpticalDuplicateFinder> finder;
+
+    public MarkDuplicatesFromShardsDataflowTransform(final PCollectionView<SAMFileHeader> header, final PCollectionView<OpticalDuplicateFinder> finder) {
+        this.header = header;
+        this.finder = finder;
+    }
+
+    @Override
+    public PCollection<GATKRead> apply(final PCollection<KV<String, Iterable<GATKRead>>> preads) {
+        return preads.apply(ParDo.named("MarkDuplicatesFromShardsDataflow")
+            .withSideInputs(header, finder)
+            .of(new DoFnWLog<KV<String, Iterable<GATKRead>>, GATKRead>("MarkDuplicatesFromShardsDataflow") {
+                private static final long serialVersionUID = 1l;
+                private transient Stopwatch processing;
+                private transient Stopwatch shipping;
+
+                @Override
+                public void startBundle(Context c) throws Exception {
+                    super.startBundle(c);
+                    processing = Stopwatch.createUnstarted();
+                    shipping = Stopwatch.createUnstarted();
+                }
+
+                @Override
+                public void finishBundle(Context c) throws Exception {
+                    super.finishBundle(c);
+                    log.info("MarkDuplicatesFromShardsDataflow bundle worked: "+processing.elapsed(TimeUnit.MILLISECONDS)+" ms. Shipped: "+shipping.elapsed(TimeUnit.MILLISECONDS)+" ms.");
+                }
+
+                // Convert from PCollection to LocalCollection, run the computation there, convert back to PCollection.
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    processing.start();
+                    final SAMFileHeader h = c.sideInput(header);
+                    final OpticalDuplicateFinder f = c.sideInput(finder);
+                    LocalCollection<GATKRead> answer = localApply(h, f, c.element().getValue());
```

Comment:
Where do you restore the header that's passed in as a side input into the read objects (since your sharded input source sometimes strips the header)? Since the `SAMRecords` are in an inconsistent state until you do this (and some operations will fail), I think it should be clearly done upfront before any local processing of the reads occurs.